### DISCUSSION TITLE
when is the data associated with a textview gets destroyed?

### DISCUSSION THREAD

**Discussion by leelavg** [frang75/nappgui_src#168] [General]
I have a textview in a layout row with text of 2.5M written (raw) increases the memory occupied by app around ~22M and when the layout row is removed memory usage doesn't go down.

I see in i_destroy_cell func a comment that says only layout/cell is removed but not the contents, is there any way to destroy the content (as per docs destroy does free as well)? Tried a couple of ways but none worked out.

I kinda have coded only 10% of my app and so didn't put it out in open yet to point to the source, pls excuse. I'm on fedora (gtk) btw.

Thanks again for the sdk. Side note, do you have any recommendations for learning MVVM & data binding, I wasn't able to find any good resources other than what's mentioned in sdk docs, pardon I'm new to both C & gui programming and so is the ask for resources for MVVM.

**Comment by frang75** [comment]
Hi @leelavg !

Thanks for report. I'll try what you mentioned about memory, just in case there is a leak on the GTK side. There are times when the memory is not released instantly. All the memory reserved by NAppGUI is managed by the Heap module, which would report any unreleased allocation.

On the other hand, there is not much documentation on Data Binding as a subject. 
All I found was the .NET documentation. 
https://learn.microsoft.com/en-us/dotnet/api/system.windows.data.binding?view=windowsdesktop-8.0

The documentation I provide at 
https://nappgui.com/en/core/dbind.html
describes my own implementation for NAppGUI. This is very useful not only for the GUI, but also for automating certain operations on program objects.

**Reply by leelavg** [reply]
Yes, I will give dbind a couple more reads. I believe memory is not being held by sdk heap module but by gtk and maybe it isn't freeing assuming it can grow again.

Anyways, thanks for the answer, will look out if there will be any update.

**Reply by leelavg** [reply]
@frang75 pardon for piggybacking on the same question as gh issues/discussions are a bit uncomfortable for a quick one off questions, let me know if you want a separate tracker.

In table after an intial draw my tableview_ondata cb just boils down to `return arrpt_get(data->ele, col + data->ncols * row, String);` + some conversion, due to this view being invoked for every mouse pos change the cpu shoots up to almost 21% for drawing ~60 rows and 3 cols.

Are there any plans to either retain data or stop invoking the func for mouse pos change or basically anything which doesn't do spurious work.

You can observe the cpu spike w/ guihello table sample as well.

Pardon if I missed anything in docs, I looked at them more than a couple of times.

**Reply by frang75** [reply]
Hi again @leelavg,

No problem where you ask the questions (Discussions or Issues). Well here there are actually two questions in one.

* Mouse movement updates the TableView area as possible highlights (rows, cells, texts, etc.) need to be updated. Currently only rows. TableView only updates the visible cells and this is done very quickly, as it involves (relatively) few calls to the low-level (very optimized) drawing API (GDI+, Cairo or Quarzt2D). If this update were really critical for some application, a flag like `tableview_OnMove_update(TableView *view, const bool_t update)` could be added. So far, I have not seen the need to do so.

* On the other hand, TableView is not expected to store the application's data in a temporary cache. It is designed as a "dumb viewer", which can display a large number of rows and columns (perhaps thousands). Its job is to calculate which portion is visible and ask the application for the data in each cell on each refresh. Any changes to the application's data will be immediately displayed in the TableView. Implementing caching in the viewer would complicate its logic. Managing these caches is the responsibility of the application, which really knows the data source. The application must ensure that the provision of the data required by the TableView is very fast. In the case that the data comes from a database or webservice (slow query), the application should keep in memory the possible data to be displayed in the TableView (perhaps a secondary thread for request and synchro the cache). More info: https://nappgui.com/en/gui/tableview.html#h2

**Reply by leelavg** [reply]
> Mouse movement updates the TableView area as possible highlights (rows, cells, texts, etc.) need to be updated
- Ack, this is the main point as due to mouse movement the visible area might be changed, understood. I know of the second point about the view being as stateless as possible and I did read the docs (they are very good btw)
- I'll see if I can patch my sdk copy in anyway that reduces CPU usage for mouse movements.

**Reply by frang75** [reply]
If you comment this line, no mouse move event will be generated.
![image](https://github.com/user-attachments/assets/280cbb7b-9aa0-4430-a756-cbc16df9afe6)

**Reply by leelavg** [reply]
Sry for not updating here, I already did that and it turns off some other functionality (like col resize etc) so I ended up commenting `view_update` in that func at the end and most of the things are working.

If anything breaks I'll make a way to this func to update view conditionally from my app.

Btw, thanks for the heap module, since I'm pedantic not to leak memory and C not being my daily driver I hooked up sdk heap to be usable by other libs which accepts a malloc and so far it's helping a lot!

Below are the patches I've so far, many of these are specific to my use case and eventually I'll be putting them up in GH if I continue to experiment w/ the app, your sdk is pleasant to experiment with 😅
```
----- nap-01-font.patch- [disregard, moved to "draw2d_preferred_monospace"]
On Fedora "Ubuntu Mono" isn't available and "Noto Sans Mono" was a combined
effort and mostly be available in many systems.

Signed-off-by: Leela Venkaiah G <leelavg@thoughtexpo.com>
----- nap-02-remove-gdk-env.patch
Remove hardcoded GDK_BACKEND value

Signed-off-by: Leela Venkaiah G <leelavg@thoughtexpo.com>
----- nap-03-wayland-std.patch
Give higher priority to user set env or fallback to XDG_SESSION_TYPE.

Current patch only calculates sizes for static windows, ie, no resize is
allowed.

Signed-off-by: Leela Venkaiah G <leelavg@thoughtexpo.com>
----- nap-04-windows-headers.patch
Windows system allows headers to be case insensitive but during cross-compile
MinGW requires these to be lowecased.

find . -type f -name 'CMakeCache.txt' -exec rm -v {} +
grep -rln '#include <Windows.h>' | xargs sed -i 's/Windows.h/windows.h/g'
grep -rln '#include <ShlObj.h>' | xargs sed -i 's/ShlObj.h/shlobj.h/'
grep -rln '#include <VersionHelpers.h>' | xargs sed -i 's/VersionHelpers.h/versionhelpers.h/'
grep -rln '#include <Commctrl.h>' | xargs sed -i 's/Commctrl.h/commctrl.h/'
grep -rln '#include <Richedit.h>' | xargs sed -i 's/Richedit.h/richedit.h/'
grep -rln '#include <WinInet.h>' | xargs sed -i 's/WinInet.h/wininet.h/'

Signed-off-by: Leela Venkaiah G <leelavg@thoughtexpo.com>
----- nap-05-mingwamd64-cross.patch
Changes require in CMake to compile windows executable on linux (fedora).

Signed-off-by: Leela Venkaiah G <leelavg@thoughtexpo.com>
----- nap-06-osxcross.patch
Use https://github.com/shepherdjerred/macos-cross-compiler to cross-compile
for MacOSX aarch64 and x86_64 on Linux

Signed-off-by: Leela Venkaiah G <leelavg@thoughtexpo.com>
----- nap-07-nonblocking-pipe.patch
Make pipe read non blocking on windows & unix

Signed-off-by: Leela Venkaiah G <leelavg@thoughtexpo.com>
----- nap-08-nowindow-cmd.patch
Do not show cmd console window when running "cmd" commands

Signed-off-by: Leela Venkaiah G <leelavg@thoughtexpo.com>
----- nap-09-proc-kill.patch
After all testing, setting sid and waitpid after kill is leaving no zombies &
orphaned process.

Signed-off-by: Leela Venkaiah G <leelavg@thoughtexpo.com>
----- nap-10-proc-wait.patch
Extra function to wait until all the pipes are closed.

Signed-off-by: Leela Venkaiah G <leelavg@thoughtexpo.com>
----- nap-11-stderr-nonblock.patch
Make stderr as well nonblocking, depends on previously made changes

Signed-off-by: Leela Venkaiah G <leelavg@thoughtexpo.com>
```

**Reply by leelavg** [reply]
reg table:
should invalidate height as well before https://github.com/frang75/nappgui_src/blob/f7b5470cbcabeb7bceb27268f718ccd118a7b9c0/src/gui/tableview.c#L1039 or else scrollbar creates empty space below the table on every header resize.

**Comment by leelavg** [comment]
Closing in favour of #176.

**Reply by leelavg** [reply]
Incorrect, tableview discussion is closed but textview memory usage still exists, as it was my mistake clubbing two asks, not reopening this.
